File: engines/unity/flutter_plugin/lib/src/models/tier_play_result.dart
/// Final tier stage state returned by the game core after one play session.
class TierPlayResult {
  const TierPlayResult({
    required this.stageIndex,
    required this.finalHealth,
    required this.maxHealth,
    required this.endingCombo,
    this.tierId,
  });

  final String? tierId;
  final int stageIndex;
  final double finalHealth;
  final double maxHealth;
  final int endingCombo;

  factory TierPlayResult.fromJson(Map<String, dynamic> json) {
    return TierPlayResult(
      tierId: json['tierId'] as String?,
      stageIndex: (json['stageIndex'] as num).toInt(),
      finalHealth: (json['finalHealth'] as num).toDouble(),
      maxHealth: (json['maxHealth'] as num).toDouble(),
      endingCombo: (json['endingCombo'] as num).toInt(),
    );
  }

  Map<String, dynamic> toJson() {
    return {
      if (tierId != null) 'tierId': tierId,
      'stageIndex': stageIndex,
      'finalHealth': finalHealth,
      'maxHealth': maxHealth,
      'endingCombo': endingCombo,
    };
  }
}

File: engines/unity/Assets/Scripts/Game/Tier/TierPlayResult.cs
using System;

[Serializable]
public class TierPlayResult
{
    public string tierId;
    public int stageIndex;
    public double finalHealth;
    public double maxHealth;
    public int endingCombo;

    public static TierPlayResult FromSession(TierPlaySession session, GameState state)
    {
        return new TierPlayResult
        {
            tierId = session.TierId,
            stageIndex = session.StageIndex,
            finalHealth = state.Health,
            maxHealth = session.MaxHealth,
            endingCombo = session.Combo
        };
    }
}
